workspaces/optic-engine/src/queries/shape.rs

```diff
@@ -438,6 +441,96 @@ impl<'a> ShapeQueries<'a> {
     let command = ShapeCommand::remove_field(field_node.field_id.clone());
     Some(std::iter::once(command))
   }
+
+  pub fn edit_shape_commands(
+    &'a self,
+    shape_trail: &ShapeTrail,
+    required_kinds: impl IntoIterator<Item = &'a ShapeKind>,
+    id_generator: &mut impl SpecIdGenerator,
+  ) -> Option<impl Iterator<Item = ShapeCommand>> {
+    let current_trail_choices = self.list_trail_choices(shape_trail);
+
+    let current_kinds: BTreeSet<_> = current_trail_choices
+      .iter()
+      .map(|choice| &choice.core_shape_kind)
+      .cloned()
+      .collect();
+
+    let togglable_kinds = if shape_trail.is_field() {
+      BTreeSet::from_iter(vec![ShapeKind::OptionalKind, ShapeKind::NullableKind])
+    } else {
+      BTreeSet::new() // only allow edits to fields for now
+    };
```

Comment:
I use `BTreeSet` instead of `HashSet` here so we can rely on the id generation and commands order being stable. That really hurt us in testing the learners, and since this is action to happen from direct user interaction, we should have the performance budget for it.